python/src/io.rs

```diff
@@ -284,13 +285,18 @@ pub struct PyIo(PyObject);
 impl Read for PyIo {
     fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
         Python::with_gil(|py| {
+            if buf.is_empty() {
+                return Ok(0);
+            }
+            let to_read = max(1, buf.len() / 4); // We divide by 4 because TextIO works with number of characters and not with number of bytes
             let read = self
                 .0
-                .call_method(py, "read", (buf.len(),), None)
+                .as_ref(py)
+                .call_method1("read", (to_read,))
```

Comment:
I understand that this is somewhat unrelated to the issue at hand, but as performance trick, I would suggest either do the method look-up once using `getattr`, storing the result in `PyIo` and using `call1` on it. Possibly using a new type distinct from `PyIo` if you want to support the `read` Python method changing during I/O processing.

Depending on how this is used, you could even have `PyIo<'py>(&'py PyAny)` to avoid the overhead of the `with_gil`, but of course, this would imply you cannot release the GIL during the whole operation.

What can also do with the current structure is to use the [`intern!`](https://docs.rs/pyo3/latest/pyo3/macro.intern.html) macro to avoid allocating a `PyString` for `"read"` for each call.

Reply:
(Since someone else was asking about this, I wonder whether we should provide such `io::Read` and `io::Write` wrappers in PyO3 directly. But the additional copying required seems to not make this universally useful, e.g. if the Python-side is backed by a file descriptor and one does not want to support sharing partially consumed buffers between Python and Rust, it is much more efficient to just pass the file descriptor to Rust and use `File::from_raw_fd`.)

Reply:
Thank you so much for the performance tricks!

- `getattr` seems a great idea, I am just a bit worried it's error prone if someone hacks around the `read` method. This method is already quite heavy so I don't know how much it's worth it if it only allows to earn a few % of performance. People reading big files should not use this code anyway but give onl the file path and let the Rust code do all the work.
- I am not sure about `PyIo<'py>(&'py PyAny)` because parsing is quite heavy and, so, releasing the GIL seems worth it.
- I have added the `intern!` calls, thank you for the trick!
- A shared `io::Read` and `io::Write` wrapper looks a great idea. It might be usable in a few places. But indeed there are a lot of "details" like text IO vs bytes IO, file descriptor optimizations just like you said...

Reply:
> getattr seems a great idea, I am just a bit worried it's error prone if someone hacks around the read method. This method is already quite heavy so I don't know how much it's worth it if it only allows to earn a few % of performance. People reading big files should not use this code anyway but give onl the file path and let the Rust code do all the work.

Indeed it is a trade-off of flexibility/dynamism versus efficiency and hence should not be implemented without measuring the gain.

IIRC the Python interpreter does more and more of this internally since attribute look-ups especially when involving method resolution order can be computationally expensive, but then again the interpreter is in a position to find out when the `read` attribute has changed so it isn't a trade-off for it. (It would be nice if there was a C API exposing this as generic notification so that we could provide such caching method slots in PyO3, but alas it is all deeply internal to CPython.)

And yes, if the underlying I/O call to going to spinning rust or even something like an S3 bucket, then the overhead of the attribute look-up will pale in comparison. With fast I/O devices like local NVMe disk, I suspect that it might be distributed fat though.